File: src/programs/Utilities/hddm/exam2.cpp
#include <fstream>
#include "hddm_x.hpp"

int report(hddm_x::HDDM &xrec);

int main()
{
   hddm_x::HDDM xrec;
   std::ofstream ofs("exam2.hddm");
   hddm_x::ostream *ostr = new hddm_x::ostream(ofs);

   //ostr->setCompression(hddm_x::k_z_compression);

   for (int n=0; n<1000000; ++n) {
      hddm_x::StudentList student = xrec.addStudents();
      student().setName("Humphrey Gaston");
      hddm_x::EnrolledList enrolled = student().addEnrolleds();
      enrolled().setYear(2005);
      enrolled().setSemester(2);
      hddm_x::CourseList course = enrolled().addCourses(3);
      course(0).setCredits(3);
      course(0).setTitle("Beginning Russian");
      course(0).addResults();
      course(0).getResult().setGrade("A-");
      course(0).getResult().setPass(true);
      course(1).setCredits(1);
      course(1).setTitle("Bohemian Poetry");
      course(1).addResults();
      course(1).getResult().setGrade("C");
      course(1).getResult().setPass(1);
      course(2).setCredits(4);
      course(2).setTitle("Developmental Psychology");
      course(2).addResults();
      course(2).getResult().setGrade("B+");
      course(2).getResult().setPass(true);
      *ostr << xrec;
      xrec.clear();
   }
   delete ostr;
   ofs.close();

   // try reading it back in from the output file just written
   std::ifstream ifs("exam2.hddm");
   hddm_x::istream istr(ifs);
   int count=0;
   while (ifs.good()) {
      istr >> xrec;
      if (count/100000*100000 == count) {
         std::cout << "event " << count << std::endl;
         report(xrec);
      }
      ++count;
   }
   std::cout << "finished after " << count << " events read." << std::endl;
   return 0;
}

int report(hddm_x::HDDM &xrec)
{
   hddm_x::CourseList courses = xrec.getCourses();
   int total_courses = courses.size();
   int total_enrolled = 0;
   int total_credits = 0;
   int total_passed = 0;
   hddm_x::CourseList::iterator iter;
   for (iter = courses.begin(); iter != courses.end(); ++iter) {
      if (iter->getResult().getPass()) {
         if (iter->getYear() > 1992) {
            total_credits += iter->getCredits();
         }
         ++total_passed;
      }
   }
   std::cout << courses().getName() << " enrolled in " 
             << total_courses << " courses "
             << "and passed " << total_passed << " of them, " << std::endl
             << "earning a total of " << total_credits
             << " credits." << std::endl;
   return 0;
}

File: src/programs/Utilities/hddm/exam2.py
import hddm_x
import sys

def report(xrec):
   courses = xrec.getCourses()
   total_courses = len(courses)
   total_enrolled = 0
   total_credits = 0
   total_passed = 0
   for course in courses:
      if course.getResult().Pass:
         if course.year > 1992:
            total_credits += course.credits
         total_passed += 1
   print courses[0].name, "enrolled in", total_courses, "courses", \
         "and passed", total_passed, "of them,"
   print "earning a total of", total_credits, "credits."

ostr = hddm_x.ostream("exam2.hddm")
#ostr.compression = hddm_x.k_z_compression

for n in range(0, 1000000):
   xrec = hddm_x.HDDM()
   student = xrec.addStudents()
   student[0].name = "Humphrey Gaston"
   enrolled = student[0].addEnrolleds()
   enrolled[0].year = 2005
   enrolled[0].semester = 2
   course = enrolled[0].addCourses(3)
   course[0].credits = 3
   course[0].title = "Beginning Russian"
   course[0].addResults()
   course[0].getResult().grade = "A-"
   course[0].getResult().Pass = True
   course[1].credits = 1
   course[1].title = "Bohemian Poetry"
   course[1].addResults()
   course[1].getResult().grade = "C"
   course[1].getResult().Pass = 1
   course[2].credits = 4
   course[2].title = "Developmental Psychology"
   course[2].addResults()
   course[2].getResult().grade = "B+"
   course[2].getResult().Pass = True
   ostr.write(xrec)
del ostr

# try reading it back in from the output file just written

istr = hddm_x.istream("exam2.hddm")
count=0
for rec in istr:
   if count % 100000 == 0:
      print "event", count
      report(rec)
   count += 1
print "finished after", count, "events read."
